pkg/node/smart.go

```diff
@@ -0,0 +1,91 @@
+/*
+Copyright © 2040 Dell Inc. or its subsidiaries. All Rights Reserved.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+   http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+// Package node contains implementation of CSI Node component
+package node
+
+import (
+	"context"
+	"sync"
+
+	api "github.com/dell/csi-baremetal/api/generated/v1"
+	smart "github.com/dell/csi-baremetal/api/smart/generated"
+	"github.com/sirupsen/logrus"
+)
+
+// SmartService represents smart API handler
+type SmartService struct {
+	mux    sync.Mutex
+	client api.DriveServiceClient
+	log    *logrus.Entry
+}
+
+// NewSmartService is the constructor for CmartService struct
+// Receives query path prefix and handles incomming HTTP requests
+// Returns an instance of SmartService
+func NewSmartService(client api.DriveServiceClient, logger *logrus.Logger) *SmartService {
+	s := &SmartService{
+		client: client,
+		log:    logger.WithField("component", "SmartService"),
+	}
+	return s
+}
+
+// GetAllDrivesSmartInfo implements get-all-drives-smart-info operation.
+//
+// Retrieve all discovered disks information/metrics.
+//
+// GET /smart
+func (s *SmartService) GetAllDrivesSmartInfo(ctx context.Context) (smart.GetAllDrivesSmartInfoRes, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	smartInfoResponse, err := s.client.GetAllDrivesSmartInfo(context.Background(), &api.Empty{})
+	if err != nil {
+		s.log.Errorf("Drivemgr response failure: %v", err)
+		return &smart.GetAllDrivesSmartInfoNotFound{}, nil
+	}
+
+	s.log.Debugf("Drivemgr response %v ", smartInfoResponse)
+
+	var smartInfo smart.OptString
+	smartInfo.SetTo(smartInfoResponse.GetSmartInfo())
+	response := smart.SmartMetrics{SmartInfo: smartInfo}
+	return &response, nil
+}
+
+// GetSmartInfo implements get-smart-info operation.
+//
+// Retrieve the disk information/metrics with the matching serial number.
+//
+// GET /smart/{serialNumber}
+func (s *SmartService) GetSmartInfo(ctx context.Context, params smart.GetSmartInfoParams) (smart.GetSmartInfoRes, error) {
+	s.mux.Lock()
+	defer s.mux.Unlock()
+
+	smartInfoResponse, err := s.client.GetSmartInfo(context.Background(), &api.SmartInfoRequest{SerialNumber: params.SerialNumber})
+	if err != nil {
+		s.log.Errorf("Drivemgr response failure: %v", err)
```

Comment:
404 seems not best for all the cases, can we somehow differentiate between not found drive for serial number and real error code?

Reply:
error handling added for 400,404,500 scenarions